llms/mlx_lm/models/gpt2.py

```diff
@@ -0,0 +1,207 @@
+        self.c_proj = nn.Linear(4 * self.n_embd, self.n_embd)
+
+    def __call__(self, x) -> mx.array:
+        return self.c_proj(nn.gelu(self.c_fc(x)))
+
+
+class TransformerBlock(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+
+        self.n_head = args.n_head
+        self.n_embd = args.n_embd
+        self.layer_norm_epsilon = args.layer_norm_epsilon
+        self.attn = Attention(args)
+        self.mlp = MLP(args)
+        self.ln_1 = nn.LayerNorm(
+            self.n_embd,
+            eps=self.layer_norm_epsilon,
+        )
+        self.ln_2 = nn.LayerNorm(self.n_embd, eps=self.layer_norm_epsilon)
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Tuple[mx.array, mx.array]] = None,
+    ) -> mx.array:
+        r = self.attn(self.ln_1(x), mask, cache)
+        h = x + r
+        r = self.mlp(self.ln_2(h))
+        out = h + r
+        return out
+
+
+class GPT2Model(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.n_embd = args.n_embd
+        self.n_positions = args.n_positions
+        self.vocab_size = args.vocab_size
+        self.n_layer = args.n_layer
+        self.layer_norm_epsilon = args.layer_norm_epsilon
+        assert self.vocab_size > 0
+        self.wte = nn.Embedding(self.vocab_size, self.n_embd)
+        self.wpe = nn.Embedding(self.n_positions, self.n_embd)
+        self.h = [TransformerBlock(args=args) for _ in range(self.n_layer)]
+        self.ln_f = nn.LayerNorm(self.n_embd, eps=self.layer_norm_epsilon)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        _, L = inputs.shape
+
+        hidden_states = self.wte(inputs)
+
+        mask = None
+        if hidden_states.shape[1] > 1:
+
+            position_ids = mx.array(np.arange(L))
+            hidden_states += self.wpe(position_ids)
+
+            mask = create_additive_causal_mask(
+                hidden_states.shape[1], cache[0].offset if cache is not None else 0
+            )
+            mask = mask.astype(hidden_states.dtype)
+
+        if cache is None:
+            cache = [None] * len(self.h)
+
+        for layer, c in zip(self.h, cache):
+            hidden_states = layer(hidden_states, mask, cache=c)
+
+        return self.ln_f(hidden_states)
+
+
+class Model(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+        self.model_type = args.model_type
+        self.model = GPT2Model(args)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        out = self.model(inputs, cache)
+        out = self.model.wte.as_linear(out)
+        return out
+
+    def sanitize(self, weights):
+        new_weights = {}
+        for i in range(self.args.n_layer):
+            if f"h.{i}.attn.bias" in weights:
+                del weights[f"h.{i}.attn.bias"]
```

Comment:
Why delete that one?

Reply:
Despite having the name "bias" it is a triangular masking matrix that was used in the original implementation. Most implementations including the HF one just use a masked_fill and create it during model initialization vs loading this weight matrix. 

https://github.com/huggingface/transformers/issues/1419#issuecomment-538505604

Reply:
Makes sense, thanks!